contracts/fee/FeeHandler.sol

```diff
@@ -3,46 +3,221 @@
+    function withdrawFees(address[] calldata markets, address buybackToken) external nonReentrant onlyFeeKeeper {
+        _validateBuybackToken(_getBatchSize(buybackToken), buybackToken);
+
+        for (uint256 i; i < markets.length; i++) {
+            FeeUtils.claimFees(dataStore, eventEmitter, markets[i], buybackToken, address(this));
         }
 
         address receiver = dataStore.getAddress(Keys.FEE_RECEIVER);
+        IERC20(buybackToken).safeTransfer(receiver, IERC20(buybackToken).balanceOf(address(this)));
+    }
+
+    // @dev claim fees in feeToken from the specified markets
+    // @param market the market from which to claim fees
+    // @param feeToken the fee tokens to claim from the market
+    function claimFees(address market, address feeToken, uint256 version) external nonReentrant {
+        if (_getBatchSize(feeToken) != 0) {
+            revert Errors.InvalidFeeToken(feeToken);
+        }
+
+        uint256 feeAmount;
+        if (version == v1) {
+            feeAmount = IVaultGovV1(vaultV1.gov()).withdrawFees(address(vaultV1), feeToken, address(this));
+        } else if (version == v2) {
+            feeAmount = FeeUtils.claimFees(dataStore, eventEmitter, market, feeToken, address(this));
+        } else {
+            revert Errors.InvalidVersion(version);
+        }
+
+        _incrementAvailableFeeAmounts(version, feeToken, feeAmount);
+    }
+
+    // @dev receive an amount in feeToken by depositing the batchSize amount of the buybackToken
+    // @param feeToken the token to receive with the fee amount calculated via an oracle price
+    // @param buybackToken the token to deposit in the amount of batchSize in return for fees
+    // @param minOutputAmount the minimum amount of the feeToken that the caller will receive
+    function buyback(
+        address feeToken,
+        address buybackToken,
+        uint256 minOutputAmount,
+        OracleUtils.SetPricesParams memory params
+    ) external nonReentrant withOraclePrices(params) {
+        if (feeToken == buybackToken) {
+            revert Errors.BuybackAndFeeTokenAreEqual(feeToken, buybackToken);
+        }
+
+        uint256 batchSize = _getBatchSize(buybackToken);
+        _validateBuybackToken(batchSize, buybackToken);
+
+        uint256 availableFeeAmount = _getAvailableFeeAmount(feeToken, buybackToken);
+        if (availableFeeAmount == 0) {
+            revert Errors.AvailableFeeAmountIsZero(feeToken, buybackToken, availableFeeAmount);
+        }
+
+        uint256 maxFeeTokenAmount = _getMaxFeeTokenAmount(feeToken, buybackToken, batchSize);
+        uint256 outputAmount = availableFeeAmount < maxFeeTokenAmount ? availableFeeAmount : maxFeeTokenAmount;
+
+        if (outputAmount < minOutputAmount) {
+            revert Errors.InsufficientBuybackOutputAmount(feeToken, buybackToken, outputAmount, minOutputAmount);
+        }
 
+        _buybackFees(feeToken, buybackToken, batchSize, outputAmount, availableFeeAmount);
```

Comment:
should emit a Buyback event here

Reply:
Ok, this has been added in the latest commit, not sure if buybackToken (and maybe feeToken) should be indexed?

Reply:
i think it is fine without index